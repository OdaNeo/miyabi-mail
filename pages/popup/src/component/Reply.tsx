import { splitByPrompts, PROMPT } from '@src/utils/tts';
import { useState } from 'react';

export function Reply({ replyOrigin }: { replyOrigin: string }) {
  const { origin, reply, result } = splitByPrompts(replyOrigin);
  const [showOrigin] = useState(true);

  return (
    <>
      <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
        <div>{PROMPT.ORIGIN}</div>
        <div>{origin}</div>
      </div>
      <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
        <div>{PROMPT.REPLY}</div>
        <div>{reply}</div>
      </div>
      <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
        <div>{PROMPT.RESULT}</div>
        <div>{result}</div>
      </div>
      {showOrigin && (
        <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
          <div>AI返回原文</div>
          <div>{replyOrigin}</div>
        </div>
      )}
    </>
  );
}
