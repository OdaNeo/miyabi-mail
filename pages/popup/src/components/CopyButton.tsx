import { Button } from '@extension/ui';
import { motion, AnimatePresence } from 'framer-motion';
import { Copy, Check } from 'lucide-react';
import { useState } from 'react';

export const CopyButton = ({ handleCopyText }: { handleCopyText: () => void }) => {
  const [isCopied, setIsCopied] = useState(false);

  const handleCopy = () => {
    if (isCopied) {
      return;
    }
    handleCopyText();
    setIsCopied(true);
    setTimeout(() => setIsCopied(false), 2000);
  };

  return (
    <Button
      variant="ghost"
      size="icon"
      data-testid="copy-icon-group"
      className="h-6 w-6 text-slate-500 hover:text-slate-900 relative"
      onClick={handleCopy}
    >
      <AnimatePresence initial={false} mode="wait">
        {isCopied ? (
          <motion.div
            key="check"
            initial={{ opacity: 0, scale: 0.8 }}
            animate={{ opacity: 1, scale: 1 }}
            exit={{ opacity: 0, scale: 0.8 }}
            transition={{ duration: 0.2 }}
            className="absolute inset-0 flex items-center justify-center"
          >
            <Check className="h-4 w-4 text-green-500" />
          </motion.div>
        ) : (
          <motion.div
            key="copy"
            initial={{ opacity: 0, scale: 0.8 }}
            animate={{ opacity: 1, scale: 1 }}
            exit={{ opacity: 0, scale: 0.8 }}
            transition={{ duration: 0.2 }}
            className="absolute inset-0 flex items-center justify-center"
          >
            <Copy className="h-4 w-4" />
          </motion.div>
        )}
      </AnimatePresence>
    </Button>
  );
};
