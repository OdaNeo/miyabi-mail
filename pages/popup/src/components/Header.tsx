import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '@src/components/ui/tooltip';
import { Button } from '@src/components/ui/button';
import { Settings, Sun, Moon, AlertTriangle, EyeOff, Eye, Copy, Trash2 } from 'lucide-react';
import { Popover, PopoverContent, PopoverTrigger } from '@src/components/ui/popover';
import { useRef, useState } from 'react';
import { Input } from '@src/components/ui/input';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from './ui/select';
import { useStorage } from '@extension/shared';
import { apiKeyStorage, apiVersionStorage, darkModeStorage } from '@extension/storage';
import type React from 'react';
import { useOpenStore } from '@src/store/openStore';

export function Header() {
  const darkMode = useStorage(darkModeStorage);
  const apiKey = useStorage(apiKeyStorage);

  const { isOpen, setIsOpen } = useOpenStore();

  const [showKey, setShowKey] = useState(false);

  const inputRef = useRef<HTMLInputElement>(null);

  const apiVersion = useStorage(apiVersionStorage);

  const handleToggleDarkMode = () => {
    document.documentElement.classList.toggle('dark');
    darkModeStorage.set(!darkMode);
  };

  const handleClosePopover = () => {
    setIsOpen(false);
  };
  const handleTogglePopover = () => {
    setIsOpen(!isOpen);
  };

  const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
    const clipboardData = e.clipboardData;
    if (clipboardData) {
      const pastedData = clipboardData.getData('text');
      apiKeyStorage.set(pastedData);
    }
  };

  const handleInputClick = () => {
    if (inputRef.current) {
      inputRef.current.select();
    }
  };

  const handleCopyKey = () => {
    navigator.clipboard.writeText(apiKey);
  };

  const handleDelete = () => {
    apiKeyStorage.set('');
  };

  return (
    <div className="flex justify-between items-center">
      <h2 className="text-base font-bold text-slate-800 dark:text-slate-100">雅返信</h2>
      <div className="flex space-x-1">
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger asChild>
              <Button
                variant="outline"
                size="icon"
                className="w-6 h-6 text-slate-600 dark:text-slate-400 hover:bg-slate-200 dark:hover:bg-slate-700"
                onClick={handleToggleDarkMode}
              >
                {darkMode ? <Sun className="h-3 w-3" /> : <Moon className="h-3 w-3" />}
              </Button>
            </TooltipTrigger>
            <TooltipContent className="bg-slate-700 text-slate-100 dark:bg-slate-200 dark:text-slate-800">
              <p>{darkMode ? '切换到亮色模式' : '切换到暗色模式'}</p>
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
        <Popover open={isOpen} onOpenChange={setIsOpen}>
          <PopoverTrigger asChild>
            <Button
              onClick={handleTogglePopover}
              variant="outline"
              size="icon"
              className="w-6 h-6 text-slate-600 dark:text-slate-400 hover:bg-slate-200 dark:hover:bg-slate-700"
            >
              <Settings className="h-3 w-3" />
            </Button>
          </PopoverTrigger>
          <PopoverContent className="w-60 bg-white dark:bg-slate-900 border border-slate-200 dark:border-slate-700">
            <div className="space-y-2">
              <h3 className="font-medium text-slate-800 dark:text-slate-200">設定 OpenAI API Key</h3>
              {!apiKey && (
                <div className="text-red-500 flex items-center text-xs">
                  <AlertTriangle className="h-3 w-3 mr-1" />
                  APIキーを設定してください。
                </div>
              )}
              <div className="relative">
                <Input
                  ref={inputRef}
                  type={showKey ? 'text' : 'password'}
                  placeholder="OpenAI API Key"
                  value={apiKey}
                  readOnly
                  onPaste={handlePaste}
                  onClick={handleInputClick}
                  className="pr-20 bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer"
                />
                <div className="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1">
                  <Button
                    variant="ghost"
                    size="icon"
                    className="h-8 w-8 text-slate-500 hover:text-slate-900"
                    onClick={() => setShowKey(!showKey)}
                  >
                    {showKey ? <EyeOff className="h-4 w-4" /> : <Eye className="h-4 w-4" />}
                  </Button>
                  {apiKey && (
                    <Button
                      variant="ghost"
                      size="icon"
                      className="h-8 w-8 text-slate-500 hover:text-slate-900"
                      onClick={handleCopyKey}
                    >
                      <Copy className="h-4 w-4" />
                    </Button>
                  )}
                </div>
              </div>
              <Select value={apiVersion} onValueChange={apiVersionStorage.set}>
                <SelectTrigger className="w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200">
                  <SelectValue placeholder="Select API Version" />
                </SelectTrigger>
                <SelectContent>
                  <SelectItem value="gpt-4o-mini">GPT-4o-mini</SelectItem>
                  <SelectItem value="gpt-4o">GPT-4o</SelectItem>
                  <SelectItem value="o1-mini">GPT-o1-mini</SelectItem>
                  <SelectItem value="o1">GPT-o1</SelectItem>
                </SelectContent>
              </Select>
              <div className="flex gap-2">
                <Button
                  disabled={!apiKey}
                  className="flex-1 bg-emerald-500 text-white hover:bg-emerald-600 disabled:bg-slate-300 dark:disabled:bg-slate-700"
                  onClick={handleClosePopover}
                >
                  保存
                </Button>
                {apiKey && (
                  <Button variant="destructive" size="icon" onClick={handleDelete}>
                    <Trash2 className="h-4 w-4" />
                  </Button>
                )}
              </div>
            </div>
          </PopoverContent>
        </Popover>
      </div>
    </div>
  );
}
