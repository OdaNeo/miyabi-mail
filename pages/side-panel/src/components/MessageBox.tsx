import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';
import type React from 'react';

export const MessageBox: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <div
      className={twMerge(
        clsx(`h-auto overflow-y-auto text-sm scrollbar-custom
           dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 
         bg-white text-slate-900 placeholder-slate-500 ${children && `p-3`}`),
      )}
    >
      {children}
    </div>
  );
};
