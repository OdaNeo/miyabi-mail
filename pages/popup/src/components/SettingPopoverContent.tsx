import { Button } from '@src/components/ui/button';
import { AlertTriangle, EyeOff, Eye, Trash2 } from 'lucide-react';
import { useRef, useState } from 'react';
import { Input } from '@src/components/ui/input';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from './ui/select';
import { useStorage } from '@extension/shared';
import { apiKeyStorage, apiVersionStorage } from '@extension/storage';
import type React from 'react';
import { useI18n } from '@src/hooks/useI18n';
import { CopyButton } from './CopyButton';

export const SettingPopoverContent = ({ setIsOpen }: { setIsOpen: (isOpen: boolean) => void }) => {
  const [showKey, setShowKey] = useState(false);

  const apiVersion = useStorage(apiVersionStorage);
  const apiKey = useStorage(apiKeyStorage);

  const inputRef = useRef<HTMLInputElement>(null);

  const { API_KEY_SETTING, PLEASE_SET_API_KEY, SAVE } = useI18n();

  const handleClosePopover = () => {
    setIsOpen(false);
  };

  const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
    const clipboardData = e.clipboardData;
    if (clipboardData) {
      const pastedData = clipboardData.getData('text');
      apiKeyStorage.set(pastedData);
    }
  };

  const handleInputClick = () => {
    if (inputRef.current) {
      inputRef.current.select();
    }
  };

  const handleCopyKey = () => {
    navigator.clipboard.writeText(apiKey);
  };

  const handleDelete = () => {
    apiKeyStorage.set('');
  };

  return (
    <div className="space-y-2">
      <h3 className="font-medium text-slate-800 dark:text-slate-200">{API_KEY_SETTING}</h3>
      {!apiKey && (
        <div className="text-red-500 flex items-center text-xs">
          <AlertTriangle className="h-3 w-3 mr-1" />
          {PLEASE_SET_API_KEY}
        </div>
      )}
      <div className="relative">
        <Input
          ref={inputRef}
          type={showKey ? 'text' : 'password'}
          placeholder="OpenAI API Key"
          value={apiKey}
          readOnly
          onPaste={handlePaste}
          onClick={handleInputClick}
          className="pr-20 bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer"
        />
        <div className="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1">
          <Button
            variant="ghost"
            size="icon"
            data-testid="eye-icon"
            className="h-8 w-8 text-slate-500 hover:text-slate-900"
            onClick={() => setShowKey(!showKey)}
          >
            {showKey ? <EyeOff className="h-4 w-4" /> : <Eye className="h-4 w-4" />}
          </Button>
          {apiKey && <CopyButton handleCopyText={handleCopyKey} />}
        </div>
      </div>
      <Select value={apiVersion} onValueChange={apiVersionStorage.set}>
        <SelectTrigger className="w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200">
          <SelectValue placeholder="Select API Version" />
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="gpt-4o-mini">GPT-4o-mini</SelectItem>
          <SelectItem value="gpt-4o">GPT-4o</SelectItem>
          <SelectItem value="o1-mini">GPT-o1-mini</SelectItem>
          <SelectItem value="o1">GPT-o1</SelectItem>
        </SelectContent>
      </Select>
      <div className="flex gap-2">
        <Button
          disabled={!apiKey}
          data-testid="save-icon"
          className="flex-1 bg-emerald-500 text-white hover:bg-emerald-600 disabled:bg-slate-300 dark:disabled:bg-slate-700"
          onClick={handleClosePopover}
        >
          {SAVE}
        </Button>
        {apiKey && (
          <Button data-testid="delete-api-icon" variant="destructive" size="icon" onClick={handleDelete}>
            <Trash2 className="h-4 w-4" />
          </Button>
        )}
      </div>
    </div>
  );
};
