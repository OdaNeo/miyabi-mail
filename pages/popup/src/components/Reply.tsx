import type { PROMPT_KEYS } from '@src/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label } from './ui/label';
import { ChevronDown, ChevronUp, ClipboardCopy } from 'lucide-react';
import { Button } from './ui/button';
import { cn } from '@src/lib/utils';
import { useState } from 'react';
import { MessageBox } from './ui/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@src/store/generatedStore';
import { useExpandedSectionStore } from '@src/store/expandedSectionStore';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const [copied, setCopied] = useState(false);

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = () => {
    navigator.clipboard
      .writeText(reply)
      .then(() => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      })
      .catch(err => {
        console.error('Failed to copy text: ', err);
      });
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="space-y-2 overflow-hidden"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section} className="space-y-1">
              <div
                className="flex justify-between items-center cursor-pointer p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                role="button"
                tabIndex={0}
              >
                <Label className="text-xs font-medium text-slate-700 dark:text-slate-300">
                  {section === 'TRANSLATION' ? '原文の翻訳：' : '日本語の返信：'}
                </Label>
                <div className="flex items-center space-x-1">
                  {section === 'REPLY' && (
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={e => {
                        e.stopPropagation();
                        handleCopy();
                      }}
                      className={cn(
                        'h-6 text-xs transition-colors',
                        copied
                          ? 'bg-emerald-500 text-white hover:bg-emerald-600'
                          : 'text-slate-600 dark:text-slate-400',
                      )}
                    >
                      {copied ? 'コピー済み' : 'コピー'}
                      <ClipboardCopy className="h-3 w-3 ml-1" />
                    </Button>
                  )}
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
