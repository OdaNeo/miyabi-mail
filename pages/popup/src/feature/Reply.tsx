import type { PROMPT_KEYS } from '@/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label, Button, cn } from '@extension/ui';
import { ChevronDown, ChevronUp, ClipboardCopy } from 'lucide-react';
import { useState } from 'react';
import { MessageBox } from '../components/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@/store/generatedStore';
import { useExpandedSectionStore } from '@/store/expandedSectionStore';
import { useI18n } from '@/hooks/useI18n';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const [copied, setCopied] = useState(false);
  const { ORIGINAL_TRANSLATION, REPLY, COPY, COPIED } = useI18n();

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = () => {
    navigator.clipboard.writeText(reply).then(() => {
      setCopied(true);
      setTimeout(() => setCopied(false), 2000);
    });
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="space-y-2 overflow-y-auto scrollbar-custom"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section} className="space-y-1">
              <div
                className="flex justify-between items-center cursor-pointer h-8 p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                role="button"
                tabIndex={0}
              >
                <Label className="font-medium text-slate-700 dark:text-slate-300">
                  {section === 'TRANSLATION' ? ORIGINAL_TRANSLATION : REPLY}
                </Label>
                <div className="flex items-center space-x-1">
                  {section === 'REPLY' && (
                    <Button
                      variant="outline"
                      size="sm"
                      data-testid="copy-button"
                      onClick={e => {
                        e.stopPropagation();
                        handleCopy();
                      }}
                      className={cn(
                        'h-6 text-sm transition-colors',
                        copied
                          ? 'bg-emerald-500 text-white hover:bg-emerald-600'
                          : 'text-slate-600 dark:text-slate-400',
                      )}
                    >
                      {copied ? COPIED : COPY}
                      <ClipboardCopy className="h-3 w-3 ml-1" />
                    </Button>
                  )}
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
