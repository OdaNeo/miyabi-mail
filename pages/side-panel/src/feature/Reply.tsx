import type { PROMPT_KEYS } from '@/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label } from '@extension/ui';
import { ChevronDown, ChevronUp } from 'lucide-react';
import { MessageBox } from '../components/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@/store/generatedStore';
import { useExpandedSectionStore } from '@/store/expandedSectionStore';
import { useI18n } from '@/hooks/useI18n';
import { CopyButton } from '@/components/CopyButton';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const { ORIGINAL_TRANSLATION, REPLY } = useI18n();

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = (section: PROMPT_KEYS) => {
    navigator.clipboard.writeText(section === 'REPLY' ? reply : translation);
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="overflow-y-auto scrollbar-custom rounded-lg border border-[#e5e7eb] dark:border-[#2e3238]
        dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 bg-white text-slate-900 placeholder-slate-500"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section}>
              <div
                className="flex justify-between items-center px-3 h-9 cursor-pointer bg-slate-100 dark:bg-[#23252b]
                 hover:bg-slate-200 dark:hover:bg-slate-800"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                data-testid={`section-${section.toLocaleLowerCase()}`}
                role="button"
                tabIndex={0}
              >
                <Label className="font-medium text-[#4b5563] dark:text-[#d1d5db]">
                  {section === 'TRANSLATION' ? ORIGINAL_TRANSLATION : REPLY}
                </Label>
                <div className="flex items-center space-x-2">
                  <CopyButton handleCopyText={() => handleCopy(section)} />
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
