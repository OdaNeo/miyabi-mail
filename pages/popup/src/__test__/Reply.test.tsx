import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { Reply } from '../components/Reply';

vi.mock('@extension/shared', () => ({
  useStorage: vi.fn().mockReturnValueOnce('This is reply text').mockReturnValueOnce('This is translation text'),
}));

const mockGeneratedStore = {
  isGenerated: true,
  setIsGenerated: vi.fn(),
};

const mockExpandedStore = {
  expandedSection: null,
  setExpandedSection: vi.fn(),
};

vi.mock('@src/store/generatedStore', () => ({
  useGeneratedStore: () => mockGeneratedStore,
}));

vi.mock('@src/store/expandedSectionStore', () => ({
  useExpandedSectionStore: () => mockExpandedStore,
}));

// Mock i18n translations
vi.mock('@src/hooks/useI18n', () => ({
  useI18n: () => ({
    ORIGINAL_TRANSLATION: 'Original Translation',
    REPLY: 'Reply',
    COPY: 'Copy',
    COPIED: 'Copied',
  }),
}));

// Mock clipboard API
const mockClipboard = {
  writeText: vi.fn().mockResolvedValue(undefined),
};

Object.assign(navigator, { clipboard: mockClipboard });

describe('Reply Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    mockGeneratedStore.isGenerated = true;
    mockExpandedStore.expandedSection = null;
  });

  it('should not render when isGenerated is false', () => {
    mockGeneratedStore.isGenerated = false;
    const { container } = render(<Reply />);
    expect(container.firstChild).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when isGenerated is true', () => {
    const { container } = render(<Reply />);
    expect(container.firstChild).not.toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('should toggle section when clicking title', async () => {
    render(<Reply />);
    const title = screen.getByText('Original Translation');

    fireEvent.click(title);

    await waitFor(() => {
      expect(mockExpandedStore.setExpandedSection).toHaveBeenCalledTimes(1);
      expect(mockExpandedStore.setExpandedSection).toHaveBeenCalledWith('TRANSLATION');
    });
  });

  it('should handle keyboard interaction for section toggle', async () => {
    render(<Reply />);
    const title = screen.getByText('Original Translation');

    fireEvent.keyUp(title, { key: 'Enter' });

    await waitFor(() => {
      expect(mockExpandedStore.setExpandedSection).toHaveBeenCalledWith('TRANSLATION');
    });
  });
});
