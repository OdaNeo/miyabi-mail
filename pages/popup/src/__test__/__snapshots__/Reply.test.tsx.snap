// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reply Component > should render correctly when generated 1`] = `
<div>
  <div
    class="space-y-2 overflow-y-auto scrollbar-custom"
    style="opacity: 0; height: 0px;"
  >
    <div
      class="space-y-1"
    >
      <div
        class="flex justify-between items-center cursor-pointer p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
        role="button"
        tabindex="0"
      >
        <label
          class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-xs font-medium text-slate-700 dark:text-slate-300"
        >
          Original Translation
        </label>
        <div
          class="flex items-center space-x-1"
        >
          <svg
            class="lucide lucide-chevron-up h-3 w-3 text-slate-500 dark:text-slate-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m18 15-6-6-6 6"
            />
          </svg>
        </div>
      </div>
      <div
        class="bg-slate-100 dark:bg-slate-800 p-1 rounded h-auto overflow-y-auto text-xs scrollbar-custom text-slate-700 dark:text-slate-300"
      >
        Translation text
      </div>
    </div>
    <div
      class="space-y-1"
    >
      <div
        class="flex justify-between items-center cursor-pointer p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
        role="button"
        tabindex="0"
      >
        <label
          class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-xs font-medium text-slate-700 dark:text-slate-300"
        >
          Reply
        </label>
        <div
          class="flex items-center space-x-1"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground rounded-md px-3 h-6 text-xs transition-colors text-slate-600 dark:text-slate-400"
            data-testid="copy-button"
          >
            Copy
            <svg
              class="lucide lucide-clipboard-copy h-3 w-3 ml-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="4"
                rx="1"
                ry="1"
                width="8"
                x="8"
                y="2"
              />
              <path
                d="M8 4H6a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2v-2"
              />
              <path
                d="M16 4h2a2 2 0 0 1 2 2v4"
              />
              <path
                d="M21 14H11"
              />
              <path
                d="m15 10-4 4 4 4"
              />
            </svg>
          </button>
          <svg
            class="lucide lucide-chevron-down h-3 w-3 text-slate-500 dark:text-slate-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
